src/sentry/incidents/subscription_processor.py

```diff
@@ -483,13 +484,12 @@ def process_update(self, subscription_update: QuerySubscriptionUpdate) -> None:
 
         if not hasattr(self, "alert_rule"):
             # QuerySubscriptions must _always_ have an associated AlertRule
-            # If the alert rule has been removed then just skip
+            # If the alert rule has been removed then clean up associated tables and return
             metrics.incr("incidents.alert_rules.no_alert_rule_for_subscription")
-            logger.error(
-                "Received an update for a subscription, but no associated alert rule exists"
-            )
-            # TODO: Delete QuerySubscription here
-            # TODO: Delete SnubaQuery here
+            delete_snuba_subscription(self.subscription)
+            query_subscription = QuerySubscription.objects.get(id=self.subscription.id)
+            SnubaQuery.objects.filter(id=query_subscription.snuba_query.id).delete()
```

Comment:
I think it would be better to build this `SnubaQuery` deletion logic into `delete_snuba_subscription`, and that can be in a separate pr. The `SnubaQuery` hanging around doesn't really hurt, other than there being a lot of orphaned rows in the table.

If you do want to build it, you should check that there are no subscriptions left related to the SnubaQuery before deleting. I think the best spot for it would be here: https://github.com/getsentry/sentry/blob/1a06bc7b19e61e5d41156f845b8321d0b63e3178/src/sentry/snuba/tasks.py#L187 after we delete the subscription

Reply:
Ok thank you! I moved the logic into that task and added a test  - the existing one handled the case where there is a related subscription left so we don't delete, then I added one for when there isn't and we do.